vernac/record.mli

```diff
@@ -36,14 +30,29 @@ val definition_structure
   -> cumulative:bool
   -> poly:bool
   -> Declarations.recursivity_kind
-  -> (coercion_flag *
-      Names.lident *
-      local_binder_expr list *
-      (local_decl_expr * record_field_attr) list *
-      Id.t * constr_expr option) list
+  -> Ast.t list
   -> GlobRef.t list
 
 val declare_existing_class : GlobRef.t -> unit
 
-(** Used by elpi *)
-val declare_structure_entry : Recordops.struc_typ -> unit
+(* Implementation internals, consult Coq developers before using;
+   current user Elpi, see https://github.com/LPCIC/coq-elpi/pull/151 *)
+module Internal : sig
```

Comment:
It would help to know why this is "internal" in your opinion.

Reply:
- `declare_projections` seems clear [not used anywhere, and indeed that is record-specific], do we agree?
- `declare_structure_entry` is similar, not used anywhere, and indeed not related to the vernacular command that does this file; in fact that is just a wrapper to record ops, but adds the record in the libobject; so it will likely go away in favor of elpi for example using `Recordops` + it's own libobjct handling?

Reply:
I've updated the comment on the corresponding commit; likely we would like to provide elpi something more principled ? Indeed how to better allow plugins to interact with internalized data is open.

Note that the idea of signalling this as `Internal` is not to prevent elpi using it, but more to warn other possible users that if you do need this, you should get in touch and discuss what your specific needs are.

Reply:
Today it's easy to plug at the interface of vernac, where terms a constr_expr.
This is not what you want to if:
- you want to provide an API that resembles a vernacular command, eg `Record...`
- you want to work with terms which are not constr_expr, but rater EConstr.t, which is what all language do (MTac, MetaCoq, ...).
Internally Coq does moves the input data from constr_expr to glob and finally EConstr, and then finally does the job of the command.
So I don't know what other projects do, but I'd like to plug at the EConstr level, but then behave exactly as `Record`, that includes declaring/generating projections, coercions and stuff.

Unfortunately these passes are not clearly separated, not even internally.
It is almost ok for constants (mostly because they are just terms) but is very entangled for inductives/records (In elpi I had to completely recode the first step for inductives for example).

So the API I'd like to see is the one where the two phases, or better 3, are exposed to the client. From "expr_entry" to "glob_entry"; from "glob_entry" to "EConstr_entry" and finally from "EConstr_entry" to roughly "Vernacstate.t -> Vernacstate.t". The `prepare_*` family of APIs is in the right direction, but does not do the steps I need exactly.

Reply:
I agree, having a better phase separation and even an interface for internalizable objects would be nice, I have been working toward thats but it's gonna be a heavy patch as likely will need to generalize some parts [that is also needed in other domains, like serapi users that would like to internalize docs etc....]

I think at least for records / inductives, we should be able to provide to elpi a much better API in subsequent PRs.